File: src/builder_base.js
/**
 * base
 */

const { KEYS, TYPES, CONFIGS } = require("./constants.js");
const { js, files, tools, httpy } = require("./tools.js");
var path = require("path");

class MyBuilderBase {

    static createFunc = null;

    constructor(buildType, ...args) {
        buildType = buildType.toLowerCase();
        this._separator = tools.get_separator();
        this._printTag = "[build_" + buildType + "_tool]";
        this._projDir = process.cwd() + this._separator;
        this._rootDir = path.join(__dirname, "..") + this._separator;
        this._workDir = path.join(this._rootDir, "build") + this._separator;
        this._buildDir = path.join(this._workDir, buildType + "_dir") + this._separator;
        this._cacheDir = path.join(this._workDir, "cache") + this._separator;
        this._libsDir = path.join(this._workDir, buildType + "_libs") + this._separator;
        files.mk_folder(this._cacheDir);
        files.mk_folder(this._libsDir);
        this._needUpdate = false;
        this._name = "UNKNOWN";
        this._isDebug = false;
        this._isRelease = false;
        this._inputNames = [];
        this._inputFiles = [];
        this._outputFile = null;
        files.mk_folder(this._buildDir);
        //
        MyBuilderBase.createFunc(this, ...args);
        console.log(`\n-----------------[JS ${buildType} Builder]---------------------\n`);
        this._print("PROJ_DIR", this._projDir);
        this._print("ROOT_DIR", this._rootDir);
        this._print("WORK_DIR", this._workDir);
    }

    _print(...args) {
        console.log(this._printTag, this._name, ...args);
    }

    _assert(v, msg) {
        if (!v) {
            throw new Error(`${this._printTag} ${this._name} ${msg}`);
        }
    }

    _error(msg) {
        throw new Error(`${this._printTag} ${this._name} ${msg}`);
    }


    _downloadByGit(url, branch, directory) {
        if (!files.is_folder(directory)) {
            this._print('cloning...');
            const cmd = `git clone ${url} ${directory} --branch ${branch} --single-branch`;
            const [isOk, err] = tools.execute(cmd);
            this._assert(isOk, "git clone failed, err:" + String(err));
        } else if (this._needUpdate) {
            this._print('pulling...');
            const cmd = `cd ${directory} && git pull`;
            const [isOk, err] = tools.execute(cmd);
            this._assert(isOk, "git pull failed:" + String(err));
        }
    }

    _downloadByTar(url, directory) {
        if (files.is_folder(directory)) {
            this._print('downloaded!');
            return;
        }
        const cacheFile = this._downloadByUrl(url);
        this._print('untarring...');
        files.mk_folder(directory);
        const cmd = `tar -xvzf ${cacheFile} -C ${directory}`;
        const [isOk, err] = tools.execute(cmd);
        files.delete(cacheFile);
        this._assert(isOk, "untar failed, err:" + String(err));
    }

    _downloadByZip(url, directory) {
        if (files.is_folder(directory)) {
            this._print('downloaded!');
            return;
        }
        const cacheFile = this._downloadByUrl(url);
        this._print('unzipping...');
        files.mk_folder(directory);
        const cmd = `unzip ${cacheFile} -d ${directory}`;
        const [isOk, err] = tools.execute(cmd);
        files.delete(cacheFile);
        this._assert(isOk, "unzip failed, err:" + String(err));
    }

    _downloadByGzip(url, directory) {
        if (files.is_folder(directory)) {
            this._print('downloaded!');
            return;
        }
        const cacheFile = this._downloadByUrl(url);
        this._print('gunzipping...');
        files.mk_folder(directory);
        const cmd = `tar xzvf ${cacheFile} -C ${directory}`;
        const [isOk, err] = tools.execute(cmd);
        files.delete(cacheFile);
        this._assert(isOk, "gunzip failed, err:" + String(err));
    }

    _downloadByUrl(url, path) {
        const parts = url.split(".");
        const ext = parts[parts.length - 1];
        const cacheFile = path || this._cacheDir + "temp." + ext;
        files.delete(cacheFile);
        this._print('downloading ...');
        const [isOk, err] = httpy.download(url, cacheFile);
        if (!isOk || files.size(cacheFile) <= 0) {
            files.delete(cacheFile);
            this._error('download failed, err:' + String(err));
        }
        this._print('download succeeded.');
        return cacheFile;
    }

    _readFile(path, isOnlyLocal, encoding = "utf-8") {
        this._print("read file:" + path);
        const isRemote = path.startsWith("http");
        this._print("is remote:" + String(isRemote));
        if (isRemote) {
            if (isOnlyLocal) {
                this._print("skip remote.");
                return;
            }
            this._print("downloading remote ...");
            path = this._downloadByUrl(this, path);
        }
        this._assert(files.is_file(path), "file not found:" + path);
        this._print("reading file ...");
        const content = files.read(path, encoding);
        if (isRemote) {
            files.delete(path);
        }
        this._assert(content.length > 0, "read file failed!, path:" + path);
        this._print("read file succeeded!");
        return content;
    }

    setName(name) {
        js.assert(js.is_text(name), 'invalid task name for builder');
        this._name = name;
    }

    getName(name) {
        return this._name;
    }

    setDebug(value) {
        js.assert(typeof value === 'boolean', 'invalid task name for builder');
        this._isDebug = value;
    }

    setRelease(value) {
        js.assert(typeof value === 'boolean', 'invalid task name for builder');
        this._isRelease = value;
    }

    setInput(...args) {
        this._print("input files ...");
        this._assert(this._inputFiles.length == 0, "input files are already defined");
        const inputArr = [...args];
        for (let i = 0; i < inputArr.length; i++) {
            let v = inputArr[i];
            let path = v;
            if (!files.is_file(path)) {
                path = this._projDir + v;
            }
            this._assert(files.is_file(path) || files.is_folder(path), "input file not found:" + v);
            this._print("input path:" + v);
            this._inputNames.push(v);
            this._inputFiles.push(path);
        }
        return this;
    }

    setOutput(path) {
        this._print("output file ...");
        this._assert(this._outputFile == null, "output file is already defined");
        this._print("output file:" + path);
        this._outputFile = this._projDir + path;
        return this;
    }

    _processBuild() {
        this._error("please implement start func ...");
    }

    start() {
        this._print('BUILD START:');
        this._processBuild();
        this._print('BUILD END!\n');
        return this;
    }

}

module.exports = {MyBuilderBase};

File: src/builder_base.py

import os
import tarfile
import zipfile

from tools import *

class MyBuilderBase:

    createFunc = None

    def __init__(self, buildType):
        buildType = buildType.lower()
        self._separator = os.path.sep
        self._printTag = f"[build_{buildType}_tool]"
        self._projDir = os.getcwd()
        self._rootDir = tools.append_path(os.path.dirname(__file__), "..")
        self._workDir = tools.append_path(self._rootDir, "build")
        self._buildDir = tools.append_path(self._workDir, f"{buildType}_dir")
        self._cacheDir = tools.append_path(self._workDir, "cache")
        self._libsDir = tools.append_path(self._workDir, f"{buildType}_libs")
        #
        print("\n\n\n\n--->", self._workDir)
        files.mk_folder(self._cacheDir)
        files.mk_folder(self._libsDir)
        self._needUpdate = False
        self._name = "UNKNOWN"
        self._isDebug = False
        self._isRelease = False
        self._inputNames = []
        self._inputFiles = []
        self._outputFile = None
        files.mk_folder(self._buildDir)
        print(f"\n-----------------[{buildType} Builder]---------------------\n")
        self._print("PROJ_DIR", self._projDir)
        self._print("ROOT_DIR", self._rootDir)
        self._print("WORK_DIR", self._workDir)

    def _parse(self, args):
        Globals.createFunc(self, args)
        

    def _print(self, *args):
        print(self._printTag, self._name, *args)

    def _assert(self, v, msg):
        if not v:
            raise Exception(f"{self._printTag} {self._name} {msg}")

    def _error(self, msg):
        raise Exception(f"{self._printTag} {self._name} {msg}")

    def _downloadByGit(self, url, branch, directory):
        if not files.is_folder(directory):
            self._print('cloning...')
            cmd = f"git clone {url} {directory} --branch {branch} --single-branch"
            isOk, err = tools.execute(cmd)
            self._assert(isOk, f"git clone failed, err: {str(err)}")
        elif self._needUpdate:
            self._print('pulling...')
            cmd = f"cd {directory} && git pull"
            isOk, err = tools.execute(cmd)
            self._assert(isOk, f"git pull failed: {str(err)}")

    def _downloadByTar(self, url, directory):
        if files.is_folder(directory):
            self._print('downloaded!')
            return
        cacheFile = self._downloadByUrl(url)
        self._print('untarring...')
        files.mk_folder(directory)
        with tarfile.open(cacheFile, "r:gz") as tar:
            tar.extractall(path=directory)
        files.delete(cacheFile)
        self._print("untar succeeded!")

    def _downloadByZip(self, url, directory):
        if files.is_folder(directory):
            self._print('downloaded!')
            return
        cacheFile = self._downloadByUrl(url)
        self._print('unzipping...')
        files.mk_folder(directory)
        with zipfile.ZipFile(cacheFile, 'r') as zip_ref:
            zip_ref.extractall(directory)
        files.delete(cacheFile)
        self._print("unzip succeeded!")

    def _downloadByGzip(self, url, directory):
        if files.is_folder(directory):
            self._print('downloaded!')
            return
        cacheFile = self._downloadByUrl(url)
        self._print('gunzipping...')
        files.mk_folder(directory)
        with tarfile.open(cacheFile, "r:gz") as tar:
            tar.extractall(path=directory)
        files.delete(cacheFile)
        self._print("gunzip succeeded!")

    def _downloadByUrl(self, url, path=None):
        parts = url.split(".")
        ext = parts[-1]
        cacheFile = path or tools.append_path(self._cacheDir, f"temp.{ext}")
        files.delete(cacheFile)
        self._print('downloading ...')
        [isOk, code, msg] = httpy.download(url, cacheFile)
        if not isOk or files.size(cacheFile) <= 0:
            files.delete(cacheFile)
            self._error(f'download failed, code:{code}, err: {msg}')
        self._print('download succeeded.')
        return cacheFile

    def _readFile(self, path, isOnlyLocal, encoding=None):
        self._print(f"read file: {path}")
        isRemote = path.startswith("http")
        self._print(f"is remote: {str(isRemote)}")
        if isRemote:
            if isOnlyLocal:
                self._print("skip remote.")
                return
            self._print("downloading remote ...")
            path = self._downloadByUrl(path)
        self._assert(files.is_file(path), f"file not found: {path}")
        self._print("reading file ...")
        content = files.read(path, encoding)
        if isRemote:
            files.delete(path)
        self._assert(len(content) > 0, f"read file failed!, path: {path}")
        self._print("read file succeeded!")
        return content

    def setName(self, name):
        py.check(py.is_text(name), 'invalid task name for builder')
        self._name = name

    def getName(self):
        return self._name

    def setDebug(self, value):
        py.check(isinstance(value, bool), 'invalid task name for builder')
        self._isDebug = value

    def setRelease(self, value):
        py.check(isinstance(value, bool), 'invalid task name for builder')
        self._isRelease = value

    def setInput(self, *args):
        self._print("input files ...")
        self._assert(len(self._inputFiles) == 0, "input files are already defined")
        inputArr = tools.as_list(args)
        for v in inputArr:
            path = v
            if not files.is_file(path):
                path = tools.append_path(self._projDir, v)
            self._assert(files.is_file(path) or files.is_folder(path), f"input file not found: {v}")
            self._print(f"input path: {v}")
            self._inputNames.append(v)
            self._inputFiles.append(path)
        return self

    def setOutput(self, path):
        self._print("output file ...")
        self._assert(self._outputFile is None, "output file is already defined")
        self._print(f"output file: {path}")
        self._outputFile = tools.append_path(self._projDir, path)
        return self

    def _processBuild(self):
        self._error("please implement build func ...")

    def start(self):
        self._print('BUILD START:')
        self._processBuild()
        self._print('BUILD END!\n')
        return self


